### ISSUE TITLE
request for including algorithm

### ISSUE THREAD

**Issue by e2002e** [hashcat/hashcat#1984]
Hello, it took me almost a year to get this working, 
https://github.com/e2002e/zhou.

I have a cool algorithm that  I called rain (because when you make the yi king's hexagrams alternate it looks like rain, and that's where I found my inspiration, and that also recalls what happens when you look at the generation of passwords with the option on).

Don't be scared with the recursion, it would work fine linearly, the idea is that you have an array of integers of the length of the password, and the integers are used to select among an array of characters. That's how my software works.
Now when the rain mode is on, for each generated word:
you add to a counter, for every character in the words, the sum of the index of the character + 1, and when the word is done, you will subtract the value of a simple function, modified according to three cases : whether the character set's length is divisible by 10, or whether it is divisible by two, or if it is odd.
And the 'simple function' is 
int sum = 0;
for(int i=1; i < value; ++i)
    sum += i;
return sum
the value being the length of the generated word.
then i add the counter's value to the array of integers, modulo by the length of the character set, and that's it.
This might give you some work but you should not regret having a new algo for word generation, I don't know if the guys at john listened to my advice but when I send then my work it was not working correctly, and since then they have added a mode that looks very much like rain (but much less powerful if you wanna known) that they called subset.
Please let me known if any questions, and refer to my code whose link I gave you above.
Thanks.

**Comment by jsteube** [comment]
Where is this used in?

**Comment by e2002e** [comment]
Is this a technical question regarding this algorithm and the link to the code i provided ? Otherwise I don't understand. But if so, it's in the generateFirst generateSecond and generateThird functions (everywhere you see a conditional testing for the value of 'rain' there is a piece of code related to that).

**Comment by roycewilliams** [comment]
He means, what software or platform uses this algorithm?

**Comment by e2002e** [comment]
None that I'm aware of, I created it I did not extract it from somewhere. Anyway how is that relevant ?

**Comment by skravets-dev** [comment]
@e2002e, it's relevant because the purpose of this software/project is to find a way to confirm and reverse cryptographic algorithms already in use. This gives the user a way to recover lost passwords for various file types. Unfortunately, while the algorithm you provided does look good and could be eloquently written, without any real-world utilization it would add additional overhead and technical debt to the project. 

However, I think it's pretty good work and quite interesting none the less!

**Comment by e2002e** [comment]
I understand that I would be a deception if this doesn't do it's job...
The only thing to be done is thoroughly testing, because it's to deep of a
thought to understand why we need special cases for character sets that are
multiple of 2 or 10.
That is why I have been constantly sending the output to `sort -u |wc` when
making tests.
Sadly, the architecture of zhou doesn't support opencl because of
recursion, and that recursion also makes any kind of serious checkpoints
feature impossible. I still will be trying to understand hashcat so that I
can implement my algorithm.
I'll give you a message on this thread if/when this is done.

**Comment by e2002e** [comment]
Hi, i have tested the code with alnum character set and it still yields correct results (no duplicates, no loss), so I conclude the algorithm name "rain" is good and have these rules:
1 When iterating on the length 'i' of the password 'pass', selecting a character in a buffer 'buf' in reference to a 'state' ; instead of using pass[i] = buf[state[i]], use pass[i] = buf[state[i+rain]%crslen] where crslen is the number of characters in the character set.

2 inside the same loop over i, add the value i + 1 to the counter we have named 'rain'.

3 we need a function that will accumulate a value along the length of the password.
int accu(int x){
  int a, b=0;
  for(int a=1; a<=x; ++a)
    b+=a;
  return b;
}
3 when the word is generated, test if the character set is a multiple of 10, if yes subtract accu(length_of_password) - 2 to the value 'rain' , if not, test if it's a multiple of 2, if yes subtract accu(length_of_password) - 4, if not subtract accu(length_of_password) - 1.
These produce an exploding effect on the generation of the words, I have piped the software using it to hashcat and used rules, it is interesting but not good because the amplifier doesn't let it go as fast as the mask and we are making many duplicates...
I've put my nose into the c file for the masks but it's too abstract for me and I miss the point where you iterate over the whole password to be able to tweak it and my test have failed. This algo is going to be included with thc-hydra's bruteforce module in later releases I hope, because this module was easy to hack. The point is markov are sweet but when you crack a length of more than 10, even on fast hashes the last character will increment slowly from a to z. So if you include 'rain' the proper way would be as an option complementary to --markov-disable,
At last I have another request/suggestion, that is to be able to loop between the --increment-min and --increment-max without having to finish the current length. Indeed it's better to have a hit on a simple password of length 14 before having tried every length 13 password.
@jsteube
